module/zfs/zio_compress.c

```diff
@@ -163,9 +163,12 @@ zio_compress_data(enum zio_compress c, abd_t *src, void *dst, size_t s_len,
 		ASSERT3U(complevel, !=, ZIO_COMPLEVEL_INHERIT);
 	}
 
+	if (*dst == NULL)
+		*dst = zio_buf_alloc(s_len);
```

Comment:
> I wonder if we could allocate d_len here, not s_len. Compression below should not write more than d_len any way, as I understand. We'd just need to return how much we allocated.

That could work.

> I am also thinking whether passing pointer to abd could be cleaner than pointer to a buffer. It would both covered size and in perspective may be could handle non-linear buffers.

That is beyond the scope of this patch and would be follow-up work. I posted this PR because I was asked to look into doing an enhancement that depends on it (different than making compression work on non-linear ABDs) and this part was ready to send ahead of the larger change.

Reply:
I just pushed a revised commit that uses `d_len` for the memory allocation.

Reply:
I see, but without allocated lendth returned, `abd = abd_get_from_buf(tmpbuf, lsize);` will be incorrect.

Reply:
From one side I see good match that we are creating zil_buf cache zones with exactly the same 1/8 step.  But from another, using buffers (usually transient) of different size may create more caching overhead.  So now I am a bit uncertain about it.  Would be cool if compression could somehow predict or grow the result buffer, but that is again dreaming. ;)